r/R/dplyr-funcs-datetime.R

```diff
@@ -372,3 +372,50 @@ binding_format_datetime <- function(x, format = "", tz = "", usetz = FALSE) {
 
   build_expr("strftime", x, options = list(format = format, locale = Sys.getlocale("LC_TIME")))
 }
+
+binding_as_date <- function(x,
+                            format = NULL,
+                            tryFormats = "%Y-%m-%d",
+                            origin = "1970-01-01",
+                            tz = "UTC",
+                            base = TRUE) {
+
+  if (is.null(format) && length(tryFormats) > 1) {
+    abort("`as.Date()` with multiple `tryFormats` is not supported in Arrow")
+  }
+
+  if (call_binding("is.Date", x)) {
+    return(x)
+
+    # cast from POSIXct
+  } else if (call_binding("is.POSIXct", x)) {
+    # base::as.Date() first converts to the desired timezone and then extracts
+    # the date, which is why we need to go through timestamp() first
+    if (base || !is.null(tz)) {
+      x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+    }
+    # POSIXct is of type double -> we need this to prevent going down the
+    # "double" branch
+    x <- x
+
+    # cast from character
+  } else if (call_binding("is.character", x)) {
+    format <- format %||% tryFormats[[1]]
+    # unit = 0L is the identifier for seconds in valid_time32_units
+    x <- build_expr("strptime", x, options = list(format = format, unit = 0L))
+
+    # cast from numeric
+  } else if (call_binding("is.numeric", x) &
+             (!call_binding("is.integer", x) | origin != "1970-01-01")) {
+    # Arrow does not support direct casting from double to date32(), but for
+    # integer-like values we can go via int32()
+    # https://issues.apache.org/jira/browse/ARROW-15798
+    # TODO revisit if arrow decides to support double -> date casting
+    x <- build_expr("cast", x, options = cast_options(to_type = int32()))
+    delta_in_sec <- call_binding("difftime", origin, "1970-01-01")
+    delta_in_sec <- build_expr("cast", delta_in_sec, options = cast_options(to_type = int64()))
+    delta_in_days <- (delta_in_sec / 86400L)$cast(int32())
+    x <- build_expr("+", x, delta_in_days)
```

Comment:
I'm a little bit confused about this section. I suspect that we might be conflating a few things together which is making this hard(er) to read than it needs to be:

* For the condition, we trigger this block if `x` is numeric and either (or both) x is not an integer or the origin is non-default. And that means we are conflating our float work around with allowing a non-default origin. I bet if we separated these two out, that would make this clearer + easier to maintain. Feel free to break this out into a separate smaller help function if that's beneficial.
* The casting on `delta_in_sec` seems off — is there a reason you want|need to case that to int64, and then later divide by something that can fit in int32, then bast it to int32? Could you start off as int32 on line 416?

Reply:
I think my thinking (for conflating numeric with origin) was that the origin `argument` only makes sense when `x` is numeric. If we had separate methods, then `origin` would only be an arg for the numeric method.

Reply:
Done. Both points addressed:
* conditions separated for `origin` different from UNIX epoch
* `delta_in_sec` is cast to `int32()` when created.

Reply:
I know re-discovered why the cast to `int64()` was needed. duration -> int32 is not supported, which means row 416 above errors.

Reply:
reverted.

Reply:
Is there a JIRA for that? We should flag it since it looks like kind of like a mistake rather than a work around to get something working.

Reply:
> I think my thinking (for conflating numeric with origin) was that the origin argument only makes sense when x is numeric. If we had separate methods, then origin would only be an arg for the numeric method.

Right, that totally makes sense that origin is only relevant for when `x` is numeric. My comment was mostly about how hard it is to reason through that if/else clause to figure out that was what was doing on. I do really think this chunk of code would be easier to reason about if we broke it up and did each type conversion on its own. We don't really need or want to do full-on method dispatch, but if we broke up each piece as a separate function it'll make what's going on in each + why much easier to follow.

Reply:
Makes sense.

Reply:
On the Jira for the cast from `duration` to `int32`. There isn't one. I'm not sure how clearly documented this is, but it seems int64 is the matching type for quite a bit of the date-time/ duration types. We could extend the scope of [ARROW-15858](https://issues.apache.org/jira/browse/ARROW-15858)? Would that make sense?

Reply:
> On the Jira for the cast from duration to int32. There isn't one. I'm not sure how clearly documented this is, but it seems int64 is the matching type for quite a bit of the date-time/ duration types. We could extend the scope of [ARROW-15858](https://issues.apache.org/jira/browse/ARROW-15858)? Would that make sense?

Right, it makes sense that we have the most coverage on int64, since that's what most of the date, time, duration types are stored as. But IMO there's no reason we shouldn't add to C++ int32 -> various types date, time duration types if int64 already works for them. 

I think ARROW-15858 is slightly separate cause it's dealing with objects of `difftime` in R (though it is related...).  

But this behavior is not great, in an ideal world, we shouldn't need to go from int32 -> int64 to then  go to a duration whenever we do these. A targeted Jira that describes that would be 💯 (and linked here so we know what we can get rid of when that's done)
```
> int32 <- Scalar$create(1, type=int32())
> int32$cast(duration())
Error: NotImplemented: Unsupported cast from int32 to duration using function cast_duration
> int64 <- Scalar$create(1, type=int64())
> int64$cast(duration())
Scalar
1
```

Tangentially: We might want to clean up the language on that ticket too, mentioning floats in there might be confusing since we already know that in Arrow going from float -> duration is not working (yet?):

```
> float <- Scalar$create(1, type=float())
> type(float)
Float32
float
> float$cast(duration())
Error: NotImplemented: Unsupported cast from float to duration using function cast_duration
```


> Makes sense. I created [ARROW-16257](https://issues.apache.org/jira/browse/ARROW-16257) for breaking it up in individual functions. I'd like to get this merged in time for 8.0.0 release and then we can follow-up to break each piece into an individual function. What do you think?

I won't demand that we do the refactoring now — but we do have to get the behavior right, and I suspect that refactoring it now will make clearer more easily what the issues are and how to fix them.

Reply:
I refactored it.

Reply:
I'm not sure how much traction we would get for enabling `int32` -> `duration`. Judging by the comments on the [time32 and time64 to duration Jira](https://issues.apache.org/jira/browse/ARROW-15996), I'd say probably not much.

Reply:
There is also [ARROW-15862](https://issues.apache.org/jira/browse/ARROW-15862) touching on the issue of conversion from numeric (integers and floats) to duration.